lib/Tools/LinearLayout.cpp

```diff
@@ -0,0 +1,382 @@
+#include "triton/Tools/StrUtil.h"
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/Support/MathExtras.h"
+
+namespace mlir::triton {
+
+namespace {
+using BasesT = LinearLayout::BasesT;
+using llvm::Twine;
+
+BasesT makeBasesMap(
+    ArrayRef<std::pair<StringAttr,
+                       ArrayRef<std::pair<StringAttr, std::vector<int32_t>>>>>
+        bases) {
+  BasesT ret;
+  for (const auto &[inDim, out] : bases) {
+    for (const auto &[outDim, basis] : out) {
+      ret[inDim][outDim] = basis;
+    }
+  }
+  return ret;
+}
+
+std::string stringifyBases(const BasesT &bases) {
+  std::string ret;
+  for (const auto &[inDim, outs] : bases) {
+    ret += " - " + inDim.str() + "\n";
+    for (const auto &[outDim, bs] : outs) {
+      ret += "   " + outDim.str() + ": [" + join(bs) + "]\n";
+    }
+  }
+  return ret;
+}
+
+BasesT validateBases(BasesT bases) {
+  if (bases.empty())
+    return bases;
+
+  for (const auto &[inDim, outs] : bases) {
+    for (const auto &[outDim, bs] : outs) {
+      if (llvm::any_of(bs, [](int32_t b) { return b < 0; })) {
+        llvm::report_fatal_error(
+            "Invalid bases passed to LinearLayout.  Expected all basis "
+            "values to be non-negative, but found a negative value for "
+            "in dimension '" +
+            Twine(inDim) + "', out dimension '" + Twine(outDim) +
+            "'.  Full list of bases:\n" + stringifyBases(bases));
+      }
+    }
+  }
+
+  // Check that the bases for each `in` dim all have the same length.
+  for (const auto &[inDim, outs] : bases) {
+    int expectedSize = outs.front().second.size();
+    for (const auto &[outDim, bs] : outs) {
+      if (bs.size() != expectedSize) {
+        llvm::report_fatal_error(
+            "Invalid bases passed to LinearLayout.  For a given `in` "
+            "dimension, expected all basis lists to have the same "
+            "length.  But this failed for in dimension '" +
+            Twine(inDim) + "'.  Full list of bases:\n" + stringifyBases(bases));
+      }
+    }
+  }
+
+  // Check that every `in` dim has the same set of `out` dims, in the same
+  // order.
+  auto expectedOuts =
+      llvm::to_vector(llvm::make_first_range(bases.front().second));
+  for (const auto &[inDim, outs] : bases) {
+    if (llvm::to_vector(llvm::make_first_range(outs)) != expectedOuts) {
+      llvm::report_fatal_error(
+          "Invalid bases passed to LinearLayout.  Expected all bases to have "
+          "the same out dimensions, in the same order.  But this failed for in "
+          "dimension '" +
+          Twine(inDim) + "'.  Full list of bases:\n" + stringifyBases(bases));
+    }
+  }
+
+  // Check surjectivity, i.e. that every `out` coordinate can be reached by some
+  // `in` coordinate.
+  //
+  // This is exponential in the number of bases.  Pretty bad!
+  //
+  // TODO(jlebar): I am told we could do better by using Gaussian elimination,
+  // e.g. with this library https://gitlab.com/hatsya/open-source/f2reduce.
```

Comment:
@lezcano Yes, all of the cases that we're interested in for Triton will easily fit within 64x64, which uses this `inplace_rref_small` codepath here...

https://gitlab.com/hatsya/open-source/f2reduce/-/blob/master/f2reduce.cpp?ref_type=heads#L419-454

...which is only 30 lines of code so I guess that it could be copied and pasted into Triton's source code rather than adding f2reduce as a dependency. I released it under the MIT licence, which is exactly the same as what Triton uses, so there should be 0 problems doing this. In general f2reduce does the thing that you suggest of performing row operations which are bitsliced (and, for large matrices, vectorised, but for linear layouts we only care about small matrices).